### PR TITLE
Scanner#checkToken(Token.ID) avoids varargs

### PR DESCRIPTION
**Pull request by schlosna**
Calls to `Scanner#checkToken(Token.ID...)` with single arguments incur expensive varargs `Object[]` array allocations on very hot parsing code paths (e.g. `org.yaml.snakeyaml.parser.ParserImpl`). 

These can be avoided in JDK 8+ by adding a new `checkToken(Token.ID)` to the `Scanner` interface with default implementation, and overriding in `ScannerImpl` to provide a more efficient implementation.

### CODE REVIEW

**Comment by asomov** [comment]
Thank you for the contribution. The migration to Java 8 should happen very soon.

File: pom.xml
Comment (asomov) [inline pom.xml:14]: this is a major change

File: src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java
Comment (asomov) [inline src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java:238]: I wonder how to measure that the code duplication provides the performance improvement.

File: pom.xml
Comment (schlosna) [reply pom.xml:14]: Understood, this will require JDK 8+ runtime. 

I'm happy to wait for JDK 8 in another PR (seems like you have the JDK 8 source/target bytecode & JDK library change in the [`Build with JDK 21` PR](https://bitbucket.org/snakeyaml/snakeyaml/pull-requests/71/diff#Lpom.xmlT13) )

From the README.md it states `SnakeYAML is a YAML 1.1 processor for the Java Virtual Machine version 8+.` since 2.0:

https://github.com/snakeyaml/snakeyaml/blob/d9e15c123efc8cb1456b7bc92e257b8c8c1d26e7/README.md?plain=1#L21

File: src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java
Comment (schlosna) [reply src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java:238]: I will try to write up a JMH benchmark and tests to quantify the before/after difference for a variety of inputs.

In ad-hoc local testing, parsing very large GiB+ yaml files via `org.yaml.snakeyaml.Yaml#parse` this change yields ~25% speedup (from > 4 minutes before to < 3 minutes after).

**Comment by schlosna** [comment]
> Thank you for the contribution. The migration to Java 8 should happen very soon.

Thanks for the quick review~ I will follow up with the JMH benchmarks, and hold off until the JDK 8 source/target changes land in origin.

File: src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java
Comment (maslovalex) [reply src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java:238]: I think this might be a major boost as there are only 11 references to the varargs version and 53 with 1 argument. It meant in all those 53 invocations the `new ID { arg }` is created :) 

I think this can be done in Java 7 sources without `default` method.
This will of course break things for those who use Scanner interface.

But I am not sure how many actually provide Custom Scanners (even in test it is used only once - CanonicalScanner).

Or we maybe can merge this after the release (which is happening soon). And the next one is planned to move to Java 8.

@asomov What do you think ?

File: pom.xml
Comment (maslovalex) [reply pom.xml:14]: I think it was an idea initially, but we rolled it back at that time... and of course forgot to fix README 🤦‍♂️ 
But maybe it is a positive sign as people (who READme) assume Java 8 requirement 🤣

File: src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java
Comment (maslovalex) [reply src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java:238]: I am inclining to the next version inclusion (because of that backward compatibility issue and close release date), even I would like to see this improvement ASAP :)

Comment (asomov) [reply src/main/java/org/yaml/snakeyaml/scanner/ScannerImpl.java:238]: Yes, I think it would be better for the next version (2.4)
Version 2.3 is almost there: https://bitbucket.org/snakeyaml/snakeyaml/wiki/Changes

**Comment by schlosna** [comment]
Added JMH `ParseBenchmark` 05ea2fae running locally on `2021 Apple MacBookPro M1 Pro` with JDK 17 & JDK 21 runtimes demonstrating ~20% to ~40% improvements on small (18.34 KiB) to medium (2.17 MiB) sized yaml documents. It might be valuable to split the JMH benchmark out to a separate PR to test baseline performance as well as use for basis of other improvements, see #14 

### VM version: JDK 17.0.11, OpenJDK 64-Bit Server VM, 17.0.11+9-LTS

## Before
```
Benchmark             (entries)  Mode  Cnt    Score     Error  Units
ParseBenchmark.load        1000  avgt    3    1.605 ±   1.170  ms/op
ParseBenchmark.load      100000  avgt    3  247.998 ±  92.043  ms/op
ParseBenchmark.parse       1000  avgt    3    1.092 ±   0.296  ms/op
ParseBenchmark.parse     100000  avgt    3  114.479 ± 177.379  ms/op
```

## After
```
Benchmark             (entries)  Mode  Cnt    Score    Error  Units
ParseBenchmark.load        1000  avgt    3    0.925 ±  0.196  ms/op
ParseBenchmark.load      100000  avgt    3  149.746 ± 61.340  ms/op
ParseBenchmark.parse       1000  avgt    3    0.602 ±  0.020  ms/op
ParseBenchmark.parse     100000  avgt    3   57.208 ±  2.728  ms/op
```


### VM version: JDK 21.0.3, OpenJDK 64-Bit Server VM, 21.0.3+9-LTS

## Before
```
Benchmark             (entries)  Mode  Cnt    Score    Error  Units
ParseBenchmark.load        1000  avgt    3    1.193 ±  0.063  ms/op
ParseBenchmark.load      100000  avgt    3  188.215 ± 13.995  ms/op
ParseBenchmark.parse       1000  avgt    3    0.895 ±  0.025  ms/op
ParseBenchmark.parse     100000  avgt    3   92.312 ±  9.998  ms/op
```

## After
```
Benchmark             (entries)  Mode  Cnt    Score    Error  Units
ParseBenchmark.load        1000  avgt    3    1.056 ±  0.074  ms/op
ParseBenchmark.load      100000  avgt    3  151.990 ± 11.969  ms/op
ParseBenchmark.parse       1000  avgt    3    0.636 ±  0.061  ms/op
ParseBenchmark.parse     100000  avgt    3   64.557 ±  1.321  ms/op
```

**Comment by schlosna** [comment]
Looking through JFR snapshots from the JMH benchmark show the `org.yaml.snakeyaml.tokens.Token$ID[]` vararg allocations from some of the single argument call sites.

image

**Comment by asomov** [comment]
@schlosna feel free to pull the latest commits with Java 8

**Review by asomov — APPROVED** [review]

**Comment by schlosna** [comment]
I think I will need to still pull and merge the bitbucket commits as they haven’t synced to GitHub yet.

**Comment by asomov** [comment]
@schlosna pushed

**Comment by asomov** [comment]
thank you!
https://bitbucket.org/snakeyaml/snakeyaml/wiki/Changes

**Pull request merged**
